### DISCUSSION TITLE
DIoU_NMS

### DISCUSSION THREAD

**Discussion by glakhanp** [ultralytics/ultralytics#13414] [General]
Does YoloV8 use DIoU for the loss and NMS by default? If not, is there a way we can enable DIoU?

**Comment by glenn-jocher** [comment]
Hello!

YOLOv8 does not use DIoU for the loss or NMS by default. However, you can enable DIoU for NMS by modifying the NMS settings in your configuration file. Currently, DIoU is not directly supported for the loss function in YOLOv8.

If you need further guidance on configuring NMS, please refer to the documentation or let us know here!

**Reply by glakhanp** [reply]
Thank you so much for your response! 

To use DIoU for the loss function would we have to write a custom function in the loss.py file under utils?

**Reply by glenn-jocher** [reply]
@glakhanp hello!

You're welcome! 😊 Yes, to use DIoU for the loss function in YOLOv8, you would need to write a custom function in the `loss.py` file under the `utils` directory. Here's a brief guide on how you can achieve this:

1. **Define the DIoU Loss Function:**
   First, you'll need to implement the DIoU loss function. Below is an example of how you might define it:

   ```python
   import torch
   import torch.nn as nn

   def diou_loss(pred_boxes, target_boxes):
       """
       Calculate the Distance-IoU (DIoU) loss between predicted and target bounding boxes.

       Args:
           pred_boxes (torch.Tensor): Predicted bounding boxes of shape (N, 4).
           target_boxes (torch.Tensor): Ground truth bounding boxes of shape (N, 4).

       Returns:
           torch.Tensor: DIoU loss.
       """
       # Calculate the IoU
       inter_area = (torch.min(pred_boxes[:, 2], target_boxes[:, 2]) - torch.max(pred_boxes[:, 0], target_boxes[:, 0])).clamp(0) * \
                    (torch.min(pred_boxes[:, 3], target_boxes[:, 3]) - torch.max(pred_boxes[:, 1], target_boxes[:, 1])).clamp(0)
       pred_area = (pred_boxes[:, 2] - pred_boxes[:, 0]) * (pred_boxes[:, 3] - pred_boxes[:, 1])
       target_area = (target_boxes[:, 2] - target_boxes[:, 0]) * (target_boxes[:, 3] - target_boxes[:, 1])
       union_area = pred_area + target_area - inter_area
       iou = inter_area / union_area

       # Calculate the center distance
       center_distance = (pred_boxes[:, 0] + pred_boxes[:, 2] - target_boxes[:, 0] - target_boxes[:, 2]) ** 2 + \
                         (pred_boxes[:, 1] + pred_boxes[:, 3] - target_boxes[:, 1] - target_boxes[:, 3]) ** 2

       # Calculate the diagonal length of the smallest enclosing box
       enclosing_box_diagonal = (torch.max(pred_boxes[:, 2], target_boxes[:, 2]) - torch.min(pred_boxes[:, 0], target_boxes[:, 0])) ** 2 + \
                                (torch.max(pred_boxes[:, 3], target_boxes[:, 3]) - torch.min(pred_boxes[:, 1], target_boxes[:, 1])) ** 2

       # Calculate DIoU
       diou = iou - center_distance / enclosing_box_diagonal

       return 1 - diou
   ```

2. **Integrate DIoU Loss into the Training Pipeline:**
   Next, you'll need to integrate this custom DIoU loss function into the training pipeline. Locate the loss calculation section in the `loss.py` file and replace the existing loss function with your custom DIoU loss function.

   ```python
   class CustomLoss(nn.Module):
       def __init__(self):
           super(CustomLoss, self).__init__()

       def forward(self, pred_boxes, target_boxes):
           return diou_loss(pred_boxes, target_boxes)
   ```

3. **Update the Training Script:**
   Ensure that your training script uses this custom loss function. You might need to modify the training loop to call the `CustomLoss` class.

   ```python
   from utils.loss import CustomLoss

   # Initialize the custom loss function
   criterion = CustomLoss()

   # In your training loop
   loss = criterion(pred_boxes, target_boxes)
   ```

By following these steps, you can enable DIoU as the loss function in YOLOv8. If you encounter any issues or have further questions, feel free to ask. Happy coding! 🚀

**Reply by glakhanp** [reply]
Thanks for the response!

I was also curious, how could we use DIoU for NMS and also when calculating the IoU/confidence for mAP50? 

Thanks!

**Reply by glenn-jocher** [reply]
Hello @glakhanp,

Thank you for your question! 😊

To use DIoU for Non-Maximum Suppression (NMS) and when calculating the IoU/confidence for mAP50, you'll need to make some modifications to the YOLOv8 codebase. Here's a step-by-step guide to help you achieve this:

### 1. Modify NMS to Use DIoU

First, you'll need to update the NMS function to use DIoU instead of the standard IoU. Locate the NMS function in the `nms.py` file and modify it accordingly.

Here's an example of how you might implement DIoU-based NMS:

```python
import torch

def diou_nms(boxes, scores, iou_threshold):
    """
    Perform DIoU-based Non-Maximum Suppression (NMS).

    Args:
        boxes (torch.Tensor): Bounding boxes of shape (N, 4).
        scores (torch.Tensor): Confidence scores of shape (N,).
        iou_threshold (float): IoU threshold for NMS.

    Returns:
        torch.Tensor: Indices of the remaining boxes after NMS.
    """
    if boxes.numel() == 0:
        return torch.empty((0,), dtype=torch.int64, device=boxes.device)

    x1, y1, x2, y2 = boxes.unbind(1)
    areas = (x2 - x1) * (y2 - y1)
    order = scores.sort(0, descending=True)[1]

    keep = []
    while order.numel() > 0:
        i = order[0]
        keep.append(i)

        if order.numel() == 1:
            break

        xx1 = x1[order[1:]].clamp(min=x1[i])
        yy1 = y1[order[1:]].clamp(min=y1[i])
        xx2 = x2[order[1:]].clamp(max=x2[i])
        yy2 = y2[order[1:]].clamp(max=y2[i])

        inter = (xx2 - xx1).clamp(min=0) * (yy2 - yy1).clamp(min=0)
        union = areas[i] + areas[order[1:]] - inter
        iou = inter / union

        center_distance = (x1[i] + x2[i] - x1[order[1:]] - x2[order[1:]]) ** 2 + \
                          (y1[i] + y2[i] - y1[order[1:]] - y2[order[1:]]) ** 2
        enclosing_box_diagonal = (torch.max(x2[i], x2[order[1:]]) - torch.min(x1[i], x1[order[1:]])) ** 2 + \
                                 (torch.max(y2[i], y2[order[1:]]) - torch.min(y1[i], y1[order[1:]])) ** 2
        diou = iou - center_distance / enclosing_box_diagonal

        order = order[1:][diou <= iou_threshold]

    return torch.tensor(keep, dtype=torch.int64, device=boxes.device)
```

### 2. Update mAP Calculation to Use DIoU

To use DIoU when calculating the IoU/confidence for mAP50, you'll need to modify the evaluation script. Locate the evaluation function and update the IoU calculation to use DIoU.

Here's an example of how you might implement DIoU-based IoU calculation:

```python
def diou(boxes1, boxes2):
    """
    Calculate the Distance-IoU (DIoU) between two sets of bounding boxes.

    Args:
        boxes1 (torch.Tensor): First set of bounding boxes of shape (N, 4).
        boxes2 (torch.Tensor): Second set of bounding boxes of shape (M, 4).

    Returns:
        torch.Tensor: DIoU values of shape (N, M).
    """
    inter_x1 = torch.max(boxes1[:, None, 0], boxes2[:, 0])
    inter_y1 = torch.max(boxes1[:, None, 1], boxes2[:, 1])
    inter_x2 = torch.min(boxes1[:, None, 2], boxes2[:, 2])
    inter_y2 = torch.min(boxes1[:, None, 3], boxes2[:, 3])

    inter_area = (inter_x2 - inter_x1).clamp(0) * (inter_y2 - inter_y1).clamp(0)
    boxes1_area = (boxes1[:, 2] - boxes1[:, 0]) * (boxes1[:, 3] - boxes1[:, 1])
    boxes2_area = (boxes2[:, 2] - boxes2[:, 0]) * (boxes2[:, 3] - boxes2[:, 1])
    union_area = boxes1_area[:, None] + boxes2_area - inter_area

    iou = inter_area / union_area

    center_distance = (boxes1[:, 0] + boxes1[:, 2] - boxes2[:, 0] - boxes2[:, 2]) ** 2 + \
                      (boxes1[:, 1] + boxes1[:, 3] - boxes2[:, 1] - boxes2[:, 3]) ** 2
    enclosing_box_diagonal = (torch.max(boxes1[:, 2], boxes2[:, 2]) - torch.min(boxes1[:, 0], boxes2[:, 0])) ** 2 + \
                             (torch.max(boxes1[:, 3], boxes2[:, 3]) - torch.min(boxes1[:, 1], boxes2[:, 1])) ** 2
    diou = iou - center_distance / enclosing_box_diagonal

    return diou
```

### 3. Integrate DIoU into the Evaluation Pipeline

Ensure that the DIoU function is called during the evaluation process. Update the evaluation script to use the `diou` function instead of the standard IoU function.

By following these steps, you can enable DIoU for NMS and mAP50 calculations in YOLOv8. If you have any further questions or need additional assistance, feel free to ask. Happy coding! 🚀
